### PR TITLE
Cache Embedding Training Pipeline: Multi-Domain LoRA Approach

### PR DESCRIPTION
**Pull request by yossiovadia**
# Cache Embedding Training Pipeline: Multi-Domain LoRA Approach

## 🎯 What This PR Solves

Semantic caching in LLM applications requires accurately distinguishing between:
- ✅ **True paraphrases** - "How to diagnose diabetes?" vs "What are diagnostic methods for diabetes?" → SHOULD cache
- ❌ **False matches** - "How to diagnose diabetes?" vs "What are symptoms of diabetes?" → SHOULD NOT cache

Generic embedding models struggle with domain-specific terminology and subtle semantic differences, leading to incorrect cache hits/misses.

**This PR provides a production-ready pipeline to train lightweight LoRA adapters** that specialize embeddings for specific domains, improving cache accuracy.

## ✅ Proven Multi-Domain Results

We validated a **single multi-domain LoRA adapter** trained on 4 domains (618K triplets total):

| Domain | Margin Improvement | Training Data | Notes |
|--------|-------------------|---------------|-------|
| **Medical** | **+14.6%*triplets | MedQuAD (44K medical Q&A) |
| **Law** | **+16.9%** | 127,739 triplets | lex_glue case law |
| **Programming** | **+11.3%** | 208,620 triplets | code_search_net docs |
| **Psychology** | **+34.9%** | ~169,000 triplets | FaithDial conversations |
| **Average** | **+19.4%** | 618,096 total | Single 582KB adapter |

**Key Benefits:**
- ✅ One 582KB adapter works across all 4 domains
- ✅ Total memory: Base model (584MB) + adapter (0.6MB) = 585MB
- ✅ No domain switching logic needed
- ✅ Simpler deployment than per-domain models

### Evaluation Methodology

**Margin = avg(positive_similarity) - avg(negative_similarity)**

This measures how well the model separates true paraphrases from related-but-different queries - the critical capability for semantic caching.

### Complete Two-Step Pipeline

**Step 1: Synthetic Data Generation** (`generate_training_data.py`)
- Uses LLM (Qwen 2.5 7B) via vLLM to generate training triplets
- 3 paraphrases + 2 hard negatives per query
- Srites, checkpoint/resume, multi-GPU support
- ~8 hours for 44K queries on 4x A10G GPUs

**Step 2: LoRA Fine-Tuning** (`lora_trainer.py`)
- Trains lightweight adapter (147K params, 0.44% of base model)
- Multiple Negatives Ranking (MNR) loss with temperature=0.05
- Single epoch training (sufficient per research)
- ~30 minutes for 130K triplets on GPU
- Output: 582KB adapter file

**AWS Automation:**
- One-command deployment to g5.12xlarge (4x A10G)
- Automated training orchestration and model download
- See `docs/QUICK_START_AWS.md` for details

## 📊 Research Foundation

Based on: **[arXiv:2504.02268v1](https://arxiv.org/pdf/2504.02268v1)** - Advancing Semantic Caching for LLMs with Domain-Specific Embeddings and Synthetic Data

Key insights:
- Small specialized models > Large general models for cache matching
- 1 epoch training sufficient with quality contrastive data
- LoRA enables efficient fine-tuning (0.44% trainable params)

### Prerequisites

You'll need unlabeled domain-specifueries in JSONL format:

```jsonl
{"query": "How to diagnose Pertussis?"}
{"query": "What are the symptoms of diabetes?"}
{"query": "What treatments are available for hypertension?"}
```

**Pre-generated training triplets:**
- All 4 domains (618K total) available at: [llm-semantic-router/semantic-router-cache-triplets](https://huggingface.co/datasets/llm-semantic-router/semantic-router-cache-triplets) (private - team access)
- Includes: medical (112K), law (127K), programming (208K), psychology (169K), and merged multi-domain triplets

**Original dataset sources (if generating your own):**
- Medical: [MedQuAD](https://github.com/abachaa/MedQuAD) (CC BY 4.0)
- Law: [lex_glue](https://huggingface.co/datasets/lex_glue)
- Programming: [code_search_net](https://huggingface.co/datasets/code_search_net)
- Psychology: [FaithDial](https://github.com/McGill-NLP/FaithDial)

### Single-Domain Training

```bash
# Step 1: Generate training triplets (~8 hours on 4x A10G)
python3 generate_training_data.py \
  --input data/cache_embeddings/medical/unlabeled_queries.jsonl \
  --output ~/cache_embedding_triplets/medical/triplets.jsonl \
  --backend vllm \
  --model Qwen/Qwen2.5-7B-Instruct \
  --paraphrases 3 \
  --negatives 2 \
  --batch-size 32 \
  --tensor-parallel 4

# Step 2: Train LoRA adapter (~30 min on GPU)
python3 lora_trainer.py \
  --train-data ~/cache_embedding_triplets/medical/triplets.jsonl \
  --base-model sentence-transformers/all-MiniLM-L12-v2 \
  --output models/medical-cache-lora \
  --epochs 1 \
  --batch-size 32 \
  --lr 2e-5 \
  --temperature 0.05

# Result: 582KB LoRA adapter ready for deployment
```

### Multi-Domain Training (Recommended)

**Important Concept**: You merge **triplets** (training data), not adapters. Each time you add a new domain, you retrain from scratch on the combined triplet dataset.

#### The Flow

```
Domain A → Generate triplets → medical_triplets.jsonl (112K)
Domain B → Generate triplets → law_triplets.jsonl (127K)
Domain C → Generate triplets → programming_triplets.jsonl (208K)
Domain D → Generate triplets → psychology_triplets.jsonl (169K)
                                    ↓
                  cat all_triplets.jsonl (merge)
                                    ↓
              multi-domain_triplets.jsonl (618K total)
                                    ↓
          Train LoRA adapter FROM SCRATCH on merged data
                                    ↓
          multi-domain-lora-adapter (582KB, works on ALL 4 domains)
```

**Key**: You don't "add to" an existing adapter. You retrain from scratch on merged triplets.

For the **+19.4% average improvement** shown above, train a single adapter across all domains:

```bash
# Step 1: Generate triplets for each domain separately (run on AWS)
python3 generate_training_data.py \
  --input data/cache_embeddings/law/unlabeled_queries.jsonl \
  --output ~/cache_embedding_triplets/law/triplets.jsonl \
  --backend vllm --model Qwen/Qwen2.5-7B-Instruct \
  --tensor-parallel 4
# ... repeat for programming, psychology, etc.

# Step 2: Merge all domain triplets
cat ~/cache_embedding_triplets/medical/triplets.jsonl \
    ~/cache_embedding_triplets/law/triplets.jsonl \
    ~/cache_embedding_triplets/programming/triplets.jsonl \
    ~/cache_embedding_triplets/psychology/triplets.jsonl \
  > ~/cache_embedding_triplets/multi-domain/triplets.jsonl

# Step 3: Train single multi-domain adapter
python3 lora_trainer.py \
  --train-data ~/cache_embedding_triplets/multi-domain/triplets.jsonl \
  --base-model sentence-transformers/all-MiniLM-L12-v2 \
  --output models/multi-domain-cache-lora \
  --epochs 1 \
  --batch-size 32 \
  --lr 2e-5

# Result: Single 582KB adapter achieving +19.4% average across 4 domains
```

**Important**: When adding new domains in the future, **always merge with existing triplets** before retraining:

```bash
# Add new biology domain to existing multi-domain model
cat ~/cache_embedding_triplets/multi-domain/triplets.jsonl \
    ~/cache_embedding_triplets/biology/triplets.jsonl \
  > ~/cache_embedding_triplets/multi-domain-v2/triplets.jsonl

# Retrain with expanded dataset
python3 lora_trainer.py \
  --train-data ~/cache_embedding_triplets/multi-domain-v2/triplets.jsonl \
  --output models/multi-domain-cache-lora-v2
```

This ensures the new adapter maintains performance on previous domains while adding the new one.

See `src/training/cache_embeddings/docs/README.md` for complete documentation.

## 📁 Key Files

| File | Purpose |
|------|---------|
| `generate_training_data.py` | **Step 1:** Creates triplets from unlabeled queries |
| `lora_trainer.py` | **Step 2:** Trains LoRA adapter with MNR loss |
| `evaluate_multi_domain.py` | Validates multi-domain performance |
| `test_lora_model.py` | Single-domain evaluation |
| `docs/README.md` | Complete technical documentation |
| `docs/index.html` | Interactive results visualization (medical domain) |

## 🔮 What's Next

This pipeline is **domain-agnostic** and ready to expand to semantic-router's other supported domains:

**Completed (4/13 domains):**
- ✅ Medical/Healthcare
- ✅ Law
- ✅ Programming  
- ✅ Psychology

**Remaining domains to train:**
- Biology
- Chemieering
- Economics
- Finance
- History
- Mathematics
- Philosophy
- Scientific

The multi-domain approach makes it practical to create a single unified cache embedding model covering all 13 domains, or train domain-specific adapters as needed.

## 🤝 Review Focus

Looking for feedback on:

1. **Multi-domain approach** - Is a single adapter for all domains preferable to per-domain models?
2. **Pipeline production-readiness** - AWS automation, error handling, checkpointing
3. **Documentation clarity** - Is the README and research docs structure clear?
4. **Integration path** - How should this integrate with semantic-router runtime?
5. **Dataset links vs inclusion** - Correct approach to reference external datasets?

## 📝 Notes

- **No dataset files included** - Links provided in README to HuggingFace datasets
- **Research docs preserved** - Detailed findings in `docs/research/` for reference

---

**Summary:** Production-ready pipeline achieving **+19.4% average improvement** across 4 domains with a sing82KB LoRA adapter. Domain-agnostic approach ready to scale to all semantic-router domains.

### CODE REVIEW

**Comment by yossiovadia** [comment]
Superseded by #1094 - clean PR with proper DCO sign-off and consolidated commit history
